.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,639 @@
+          fi
+
+      - name: Enable signed images
+        if: inputs.sign-image == true && inputs.sign-api-url != ''
+        env:
+          SIGN_API: "${{ inputs.sign-api-url }}"
+          SIGN_API_KEY: "${{ secrets.SIGN_API_KEY }}"
+          SIGN_GRUB_KEY_ID: 2EB29B4CE0132F6337897F5FB8A88D1C62FCC729
+          SIGN_KMOD_KEY_APPEND: "${{ secrets.SIGN_KMOD_KEY_APPEND }}"
+        run: |
+          echo "BARYS_ARGUMENTS_VAR=${BARYS_ARGUMENTS_VAR} -a SIGN_API=${SIGN_API} -a SIGN_API_KEY=${SIGN_API_KEY} -a SIGN_GRUB_KEY_ID=${SIGN_GRUB_KEY_ID} -a SIGN_KMOD_KEY_APPEND=${SIGN_KMOD_KEY_APPEND} --bb-args --no-setscene" >> $GITHUB_ENV
+
+      - name: Mount shared cache
+        if: inputs.yocto-cache-host != '' && contains(fromJSON(inputs.runs-on), 'self-hosted')
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y --no-install-recommends nfs-common
+          sudo mount -t nfs "${{ inputs.yocto-cache-host }}:/" "${YOCTO_CACHE_DIR}" -o fsc
+
+      - name: Build
+        id: build
+        run: |
+          ./balena-yocto-scripts/build/balena-build.sh -d "${MACHINE}" -t "${{ secrets.BALENA_API_KEY }}" -s "${YOCTO_CACHE_DIR}" -g "${BARYS_ARGUMENTS_VAR}"
+          hostapp_path=$(find "${WORKSPACE}/build/tmp/deploy/" -name "balena-image-${MACHINE}.docker" -type l || true)
+          echo "hostapp_path=${hostapp_path}" >> $GITHUB_OUTPUT
+
+      - name: Prepare artifacts
+        id: prepare
+        run: |
+          source "${automation_dir}/include/balena-deploy.inc"
+          balena_deploy_artifacts "${{ inputs.machine }}" "${WORKSPACE}/gh-deploy" false
+          tar --auto-compress -cvf "${{ runner.temp }}/build-artifacts.tar.zst" "${WORKSPACE}/gh-deploy"
+
+      # https://github.com/actions/upload-artifact
+      - name: Upload artifacts
+        uses: actions/upload-artifact@26f96dfa697d77e81fd5907df203aa23a56210a8 # v4.3.0
+        with:
+          name: build-artifacts
+          path: ${{ runner.temp }}/build-artifacts.tar.zst
+          retention-days: 3
+
+  s3-deploy:
+    name: S3 Deploy
+    runs-on: ${{ fromJSON(inputs.runs-on) }}
+    needs: build
+    if: inputs.deploy-s3 == true
+
+    env:
+      MACHINE: "${{ inputs.machine }}"
+      BALENARC_BALENA_URL: "${{ inputs.balena-environment }}"
+      WORKSPACE: ${{ github.workspace }}
+      VERBOSE: verbose
+
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+
+    steps:
+      # TODO: clone balena-yocto-scripts @ yocto_scripts_ref instead
+      # https://github.com/actions/checkout
+      - name: Clone device repository
+        uses: actions/checkout@v3
+        with:
+          repository: ${{ inputs.device-repo || github.repository }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          ref: ${{ inputs.device-repo-ref || github.ref }}
+          submodules: true
+
+      - name: Update balena-yocto-scripts submodule to ${{ inputs.yocto-scripts-ref }}
```

Comment:
We could also avoid this update as we have done it before.

Reply:
See above answer, we are using 4 separate jobs to make use of parallelism and each job is a fresh workspace.

If this turns out to not be faster, we can revert to one job and workspace.